storage/mega/storage_mega.go

```diff
@@ -0,0 +1,277 @@
+/*
+ * knoxite
+ *     Copyright (c) 2016-2020, Christian Muehlhaeuser <muesli@gmail.com>
+ *     Copyright (c) 2020, Nicolas Martin <penguwin@penguwin.eu>
+ *     Copyright (c) 2020, Matthias Hartmann <mahartma@mahartma.com>
+ *
+ *   For license see LICENSE
+ */
+
+package mega
+
+import (
+	"errors"
+	"net/url"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/t3rm1n4l/go-mega"
+
+	knoxite "github.com/knoxite/knoxite/lib"
+)
+
+// StorageMega stores data on a remote Mega
+type StorageMega struct {
+	url  url.URL
+	mega *mega.Mega
+	knoxite.StorageFilesystem
+}
+
+func init() {
+	knoxite.RegisterBackendFactory(&StorageMega{})
+}
+
+// NewBackend returns a StorageMega backend
+func (*StorageMega) NewBackend(u url.URL) (knoxite.Backend, error) {
+	backend := StorageMega{
+		url:  u,
+		mega: mega.New(),
+	}
+
+	// checking for username and password
+	if u.User == nil || u.User.Username() == "" {
+		return &StorageMega{}, knoxite.ErrInvalidUsername
+	}
+	pw, pwexist := u.User.Password()
+	if !pwexist {
+		return &StorageMega{}, knoxite.ErrInvalidPassword
+	}
+
+	err := backend.mega.Login(u.User.Username(), pw)
+	// log into the mega client for accessing the API
+	if err != nil {
+		return &StorageMega{}, err
+	}
+
+	knfs, err := knoxite.NewStorageFilesystem(u.Path, &backend)
+	if err != nil {
+		return &StorageMega{}, err
+	}
+	backend.StorageFilesystem = knfs
+	return &backend, nil
+}
+
+// Location returns the type and location of the repository
+func (backend *StorageMega) Location() string {
+	return backend.url.String()
+}
+
+// Close the backend
+func (backend *StorageMega) Close() error {
+	return nil
+}
+
+// Protocols returns the Protocol Schemes supported by this backend
+func (backend *StorageMega) Protocols() []string {
+	return []string{"mega"}
+}
+
+// Description returns a user-friendly description for this backend
+func (backend *StorageMega) Description() string {
+	return "mega.nz storage"
+}
+
+// AvailableSpace returns the free space on this backend
+func (backend *StorageMega) AvailableSpace() (uint64, error) {
+	quota, err := backend.mega.GetQuota()
+	if err != nil {
+		return 0, knoxite.ErrAvailableSpaceUnknown
+	}
+	return quota.Mstrg, nil
```

Comment:
`quota.Mstrg` returns the total storage capacity of your Mega account. We need to subtract `quota.Cstrg` from it to calculate the actual available free space.